rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,587 @@
+
+    size_t actual_read_size = fread(buf, sizeof(unsigned char), read_size, _current_fPtr);
+    return actual_read_size;
+}
+
+int NumpyDataReader::close() {
+    return release();
+}
+
+NumpyDataReader::~NumpyDataReader() {
+    release();
+}
+
+int NumpyDataReader::release() {
+    if (!_current_fPtr)
+        return 0;
+    fclose(_current_fPtr);
+    _current_fPtr = nullptr;
+    return 0;
+}
+
+void NumpyDataReader::reset() {
+    if (_shuffle) {
+        std::mt19937 rng(_seed);
+        std::shuffle(_file_names.begin() + _shard_start_idx_vector[_shard_id],
+                     _file_names.begin() + _shard_start_idx_vector[_shard_id] + actual_shard_size_without_padding(), rng);
+    }
+
+    if (_stick_to_shard == false)  // Pick elements from the next shard - hence increment shard_id
+        increment_shard_id();      // Should work for both single and multiple shards
+
+    _read_counter = 0;
+
+    if (_sharding_info.last_batch_policy == RocalBatchPolicy::DROP) {  // Skipping the dropped batch in next epoch
+        for (uint i = 0; i < _batch_size; i++)
+            increment_curr_file_idx(_file_names.size());
+    }
+}
+
+Reader::Status NumpyDataReader::generate_file_names() {
+    if ((_sub_dir = opendir(_folder_path.c_str())) == nullptr)
+        THROW("NumpyDataReader ShardID [" + TOSTR(_shard_id) + "] ERROR: Failed opening the directory at " + _folder_path);
+
+    std::vector<std::string> entry_name_list;
+    std::string full_path = _folder_path;
+
+    while ((_entity = readdir(_sub_dir)) != nullptr) {
+        std::string entry_name(_entity->d_name);
+        if (strcmp(_entity->d_name, ".") == 0 || strcmp(_entity->d_name, "..") == 0) continue;
+        entry_name_list.push_back(entry_name);
+    }
+    closedir(_sub_dir);
+    std::sort(entry_name_list.begin(), entry_name_list.end());
+
+    auto ret = Reader::Status::OK;
+    if (!_file_list_path.empty()) {  // Reads the file paths from the file list and adds to file_names vector for decoding
+        if (_meta_data_reader) {
+            auto vec_rel_file_path = _meta_data_reader->get_relative_file_path();  // Get the relative file path's from meta_data_reader
+            for (auto file_path : vec_rel_file_path) {
+                if (filesys::path(file_path).is_relative()) {  // Only add root path if the file list contains relative file paths
+                    if (!filesys::exists(_folder_path))
+                        THROW("File list contains relative paths but root path doesn't exists");
+                    _absolute_file_path = _folder_path + "/" + file_path;
+                }
+                if (filesys::is_regular_file(_absolute_file_path)) {
+                    _file_names.push_back(_absolute_file_path);
+                    _file_count_all_shards++;
+                }
+            }
+            _last_file_name = _absolute_file_path;
+        } else {
+            std::ifstream fp(_file_list_path);
```

Comment:
Will we have a list file passed to numpy reader?
Do we have such use case?

Reply:
For trainings, we only pass a list of file paths to numpy reader. But DALI has the option to pass file lists and since we already support it in file_source_reader, I have added it for numpy reader as well